tests/test/snapshot/test_snapshot_client_server.cpp

```diff
@@ -139,41 +140,71 @@ TEST_CASE_METHOD(SnapshotClientServerFixture,
 
     setUpFunctionGroup(appId, groupIdB);
 
-    // Set up a snapshot
+    // Set up a snapshot that's got enough memory to expand into
     std::string snapKey = std::to_string(generateGid());
-    SnapshotData snap = takeSnapshot(snapKey, 5, true);
+    size_t initialSnapSize = 5 * HOST_PAGE_SIZE;
+    size_t expandedSnapSize = 10 * HOST_PAGE_SIZE;
+
+    faabric::util::SnapshotData snap;
+    snap.size = initialSnapSize;
+    snap.data = faabric::util::allocateVirtualMemory(expandedSnapSize);
```

Comment:
Could you introduce a std::unique_ptr-like (or even just std::unique_ptr with a custom deleter) wrapper for these allocations? I've been purging all uses of new/delete and using RAII memory management where possible, this feels like a step backwards. I know this could be impossible for the general case of applying snapshots onto faaslets, but at least in testing code the lifetime of these mappings is limited to a function call scope.

Reply:
Two observations after a bit of fiddling:

- To interface properly with Faasm, the snapshot API still needs to support raw pointers (as both WAMR and WAVM still use them for wasm memory), hence we can't go full RAII on a lot of the changes introduced in this PR. We could introduce a parallel API that supports smart pointers alongside raw pointers, but this will result in a fair bit of extra code.

- I agree that using RAII in tests for memory provisioned on the fly is a good idea, and I can almost get something with a `unique_ptr` working, however, the custom deleter needs to call `munmap` and needs to know the size of the memory that was orinally `mmap`ed, which makes defining the delete a little more tricky. I got this far:

```
std::unique_ptr<uint8_t[]> doAlloc(size_t size, int prot, int flags)
{
    auto deleter = [size](uint8_t* u) { munmap(u, size); };
    std::unique_ptr<uint8_t[], decltype(deleter)> mem(
        (uint8_t*)::mmap(nullptr, size, prot, flags, -1, 0), deleter);
    return mem;
}
```

However, this won't compile as the return type of the function also needs to specify the type of the deleter, but I can't work out the syntax. Any ideas?

Reply:
You can use `std::function<void(uint8_t*)>` as the deleter type to store that lambda

Reply:
The alternative is to use a `shared_ptr`, which I think should be more flexible re. the custom deleter.

Reply:
Here's a minimal working example with unique_ptr:
```
#include <memory>
#include <functional>
#include <sys/mman.h>

using OwnedMmapRegion = std::unique_ptr<uint8_t[], std::function<void(uint8_t*)>>;

OwnedMmapRegion doAlloc(size_t size, int prot, int flags)
{
    auto deleter = [size](uint8_t* u) { munmap(u, size); };
    OwnedMmapRegion mem(
        (uint8_t*)::mmap(nullptr, size, prot, flags, -1, 0), deleter);
    return mem;
}
```

<https://godbolt.org/z/TdTzaz9rW>

I'm not sure what you mean by shared_ptr being more flexible with the custom deleter, the only difference would be that you could use reference counting to keep track when the last user is gone.

Reply:
Awesome, cheers @kubasz. What I meant re. `shared_ptr` being more flexible is that I don't think the deleter type needs to be included in the pointer type, unlike a `unique_ptr`, so the return type of the function could still be `std::shared_ptr<uint8_t[]>` even with the deleter (and I could dodge the original problem I was facing). Anywho, your version is great so I'll see how things turn out with that.